### PR TITLE
Change UserList view from Rest APIView to Django DetailView

### PR DESCRIPTION
**Pull request by kbeker**
I have change view to django generic view, but I have one business question: Only Admin should be able get employees list? @Karrp

### CODE REVIEW

File: users/tests/test_unit_users_apiview.py [hunk 1]
```diff
@@ -12,26 +12,29 @@
 class UserListTests(TestCase):
     def setUp(self):
         self.user = CustomUser(
-            email="testuser@codepoets.it", password="newuserpasswd", first_name="John", last_name="Doe", country="PL"
+            email="testuser@codepoets.it",
```
Comment (rwrzesien) [inline users/tests/test_unit_users_apiview.py]: I see it was there before, but I think you can now use safely values generated by factory here, so you can remove `email`, `first_name` and `last_name`

Comment (rwrzesien) [reply users/tests/test_unit_users_apiview.py]: If this tests Django views please move this tests to `test_views.py`, we will then clearly see how much is left to change :)

File: users/views.py [hunk 2]
```diff
@@ -199,23 +204,19 @@ def post(cls, request: HttpRequest, pk: int) -> Union[Response, HttpResponseRedi
...
-        users_queryset = cls.get_queryset()
-        users_serializer = UserListSerializer(context={"request": request})
-        return Response({"serializer": users_serializer, "users_list": users_queryset})
+    model = CustomUser
+    queryset = CustomUser.objects.prefetch_related("projects")
+
+    def get_context_data(
+        self, *, object_list: Optional[List] = None, **kwargs: Any  # pylint: disable=unused-argument
+    ) -> dict:
+        if self.request.user.is_admin:
```
Comment (rwrzesien) [inline users/views.py]: Nice try ;) But lets add a decorator for it that we can reuse all over the place. This decorator should take a list of user types that are allowed for a given view, this way we have only one decorator for everything!

File: users/tests/test_unit_users_apiview.py [hunk 3]
```diff
@@ -12,26 +12,29 @@
...
+        response = self.client.get(self.url)
         self.assertEqual(response.status_code, 200)
         self.assertContains(response, self.user.user_type)
-        users_list = response.data["users_list"]
-        self.assertTrue(self.user in users_list)
+        self.assertContains(response, self.user.email)
+        self.assertContains(response, self.user.first_name)
+        self.assertContains(response, self.user.last_name)
 
     def test_user_list_view_should_not_be_accessible_for_unauthenticated_user(self):
```
Comment (rwrzesien) [inline users/tests/test_unit_users_apiview.py]: How about adding a test that non-admin users can't access the view? There is already similar one called `test_delete_project_function_view_should_not_delete_project_on_non_admin_request`.

File: utils/decorators.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+from functools import wraps
+from typing import Callable
+from typing import Tuple
+
+from django.contrib.auth.views import redirect_to_login
+from django.core.handlers.wsgi import WSGIRequest
+
+
+def check_permissions(permissions: Tuple[str], redirect_path: str = "/accounts/login/") -> Callable:
```
Comment (rwrzesien) [inline utils/decorators.py]: "/accounts/login/" => Is it possible to use `reverse` here?

File: utils/decorators.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    def decorator(view_func: Callable) -> Callable:
+        @wraps(view_func)
+        def _wrapped_view(request: WSGIRequest, *args: dict, **kwargs: dict) -> Callable:
+            for permission in permissions:
+                if not getattr(request.user, permission):
```
Comment (rwrzesien) [inline utils/decorators.py]: I would base it on database field `user_type`, those properties like `is_admin` might change in the future.

File: utils/decorators.py (hunk 4 above)
Comment (kbeker) [reply utils/decorators.py]: I tried at first, but It does not working in this case. I had to left it as it is.

File: utils/decorators.py (hunk 5 above)
Comment (kbeker) [reply utils/decorators.py]: Changed. Now we checking `user_type`

File: utils/decorators.py (hunk 4 above)
Comment (rwrzesien) [reply utils/decorators.py]: Then please try to set it to `None` and then in decorator code `if None then reverse()`. Should work I think.

File: utils/decorators.py (hunk 5 above)
Comment (rwrzesien) [reply utils/decorators.py]: Looks good, only thing I would like to ask you to add is an assert that passed `allowed_user_types` are correct.

Comment (kbeker) [reply utils/decorators.py]: OK, In progress

**Comment by Karrp** [comment]
Yes, only the administrator should be able to obtain a list of employees.

**Pull request merged**